### DISCUSSION TITLE
Inline properties reference

### DISCUSSION THREAD

**Discussion by stdword** [stdword/logseq13-full-house-plugin#34] [Showroom]
<img width="800px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/d360c4a6-149b-449e-ae7f-2c68979ac074"/>

## Steps

1) Create `🏛view`  (use copy button in the top right corner):
   ````markdown
   - template:: reference property
     template-version:: 1
     plugin-version:: 3.3.0
     link:: [Inline properties reference](https://github.com/stdword/logseq13-full-house-plugin/discussions/34)
       - ```typescript
         ``{
           var name = c.args.$1
           if (!name)
             return ''
           
           var newValue = undefined
           if (name.indexOf(' ') !== -1) {
             [name, ...newValue] = name.split(' ')
             newValue = newValue.join(' ')
           }
         
           var value = c.block.props[name]
           if (!newValue && value === undefined)
             return ''
           
           if (newValue && value !== newValue)
             top.logseq.api.upsert_block_property(c.block.uuid, name, newValue)
         _}``
         
         ``value``
         ```
   ````
2) Create `:macros` in `config.edn`:
   ```clojure
   :macros {
      "p" "{{renderer :template-view, reference property, $1}}"
   }
   ```
3) Restart Logseq app
4) Use `p` macro to reference properties as in demo above.

## Additional info
- Logseq v0.10.0
- ☕️ [BuJo Theme](https://github.com/stdword/logseq-bujo-theme)
- → [discuss.logseq.com](https://discuss.logseq.com/t/inline-properties-and-reuse-of-property-values-basic-implementation/23363/2?u=stdword)

**Comment by rowsof** [comment]
This is really awesome! I'm using it in a sort of inverted way where I define a bunch of properties in a block manually and then construct sentences that reference those properties. My question is: Is there any way to search properties in all blocks on the page, rather than just the current block? I'm a mildly technically savvy non-programmer. Thanks for the plugin and any help!

**Reply by stdword** [reply]
Hello!
Sure, it is possible to get properties from all blocks of the page. But the question is: which block should we prefer when several blocks has the same property? A top / bottom one? A level 1 or deep level?

Right now page properties could be accessed very easily with the help of 🏛️`inline views`:
<img width="350px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/503c4b0a-b570-432a-af29-af2fc4511062"/>

What task are you trying to solve?

**Reply by rowsof** [reply]
So, I'm a fantasy writer and I'm setting up some templates for different categories of worldbuilding. This means that on each page there will be MANY properties (too many, probably), but they will all be unique. I suppose I'd love to be able to either say {{all.p propertyName}} to find all properties in all blocks in the page, or use syntax like {{blockName.p propertyName}} to identify which block my property will be in? Anything along those lines possible? I've poked at the script you shared above and it is quite beyond me lol.

![d19bf1a6-4ee4-45dc-98b6-84ff5d783269](https://github.com/stdword/logseq13-full-house-plugin/assets/138329522/7ae4a4ea-2371-4bdc-b799-d5a0e35230fd)

Pasting the main block of text from the gif, in case it's annoying to try to read there:

`{{renderer :view, c.page.name}} is a {{p gender}} {{p race}} born on {{p birthdate}} in {{p birthplace}}. During story, {{renderer :view, c.page.name}} is {{p age-in-story}} years old. Thus far these attributes have been from either the page or the current block. But what if I wanted to access something from a separate block, like {{p dna-primary}}.`

Thanks in advance!

**Reply by stdword** [reply]
Here is my thougths:

## `{{all.p propertyName}}`
This is totally achievable with an respect to uniqueness of properties names in the page.

But uniqueness is not a restriction. We can use layers to prioritize blocks to search properties in: current block, page's first block, pages other 1-level blocks top to bottom, page's 2-level blocks top to bottom. It solves your example case with «I want to access this!». If we cannot find property in one layer, it starts to find in the next one.

## `{{blockName.p propertyName}}`
This is achievable too, but such approach has disadvantage: if you change block name (title), then all these `blockName.p` calls become wrong. They won't be auto-corrected.

### How to overcome this?
#### via block UUID
Instead we can use something like ``{{block.p ((...uuid...)) propertyName}}`` to be protected from changing block title. But this way it is not such an easy-to-type approach.

#### via sub-pages
It requires you to change page structure.
Instead of:
<img width="377" alt="CleanShot 2024-02-06 at 17 53 04@2x" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/c5e9787b-5ea1-4d41-bed4-ec3162c95475">
Use:
<img width="600px" src="https://github.com/stdword/logseq13-full-house-plugin/assets/1984175/28d29fad-2ee1-49fe-9ab3-3e19863155b5"/>

<s>Sub-pages will be renamed automatically. But you need to keep pages structure.</s>

**UPDATE**: I was pretty sure that macro call with page reference in the form ``[[...]]`` will be renamed automatically:
`{{renderer :view, "c.page.props.primary", :page [[CHARACTER TEMPLATE/Character DNA]]}}`
but It doesn't work =( so sad. Just left the comment in [corresponding bug](https://github.com/logseq/logseq/issues/6230).

Well, sub-pages is not an option. The rest is reasonable. What do you think?

**Reply by rowsof** [reply]
Awesome response, thank you for so much consideration!

I agree about your sentiments on searching by block title; it'd likely need too much maintenance.

One of my main constraints with the block UUID and the sub-page solution is that there will be a handful of character pages that use this template, and each character will have unique information.
- Am I right to assume the UUID will need to be set each time I use this template to create a new page for a character?
- Similar issue for sub-pages right? I would need to create sub-pages for each section for each character?

If the answers to the above questions are "yes," it seems in my case it would make sense to just search all the properties in all blocks on the page, as in that first example {{all.p propertyName}}. I still have no idea how to do this one 😅

**Reply by stdword** [reply]
> If the answers to the above questions are "yes"

Yes, it is. So the chosen approach is `{{all.p propertyName}}`

> I still have no idea how to do this one 😅

Yep, I just want to clarify the details before any coding.
I need some time to dive into this.

**Reply by rowsof** [reply]
Thank you! I really appreciate you spending time and effort on this. I'll take the help whenever you can get around to it! ❤️

**Reply by stdword** [reply]
Hello, @rowsof !
Just implemented this idea in new version of plugin (v3.4.0).
Setup instructions are [here](https://github.com/stdword/logseq13-full-house-plugin/discussions/37).
